src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1120 @@
+    //   }
+    //
+    // The second form contains essentially the same data, but in a more
+    // "compressed" form. In this form, the shadow block is stored as a simple
+    // [type, value] pair, where the type is a constant representing the opcode
+    // and field name in which the value should be placed when deserializing.
+    // Because it's so much more concise than the non-compressed form, most
+    // inputs are serialized in this way. The same input above in compressed
+    // form would look like this:
+    //
+    //   [4, 50]
+    //
+    // As described in serializeInputsToInputs, compressed shadow blocks are
+    // stored inline with the input they correspond to, not as separate blocks
+    // with IDs.
+    //
+    // Within this code, we use the primimtiveOrOpCode option to determine how
+    // the shadow should be serialized. If it is a number, it's referring to a
+    // "primitive", the term uses for shadows when they are in the compressed
+    // form. If it's a string, it is a (shadow) block opcode, and should be
+    // serialized in the expanded form.
+
+    const { blockData, getBroadcastId, parentId, primitiveOrOpCode } = options;
+
+    let shadowValue = null;
+
+    if (primitiveOrOpCode === BIS.BROADCAST_PRIMITIVE) {
+      // Broadcast primitives, unlike all other primitives, expect two values:
+      // the broadcast name and its ID.
+      shadowValue = [BIS.BROADCAST_PRIMITIVE, value, getBroadcastId(value)];
+    } else if (primitiveOrOpCode === BIS.COLOR_PICKER_PRIMITIVE) {
+      // Color primitive. Convert the {r, g, b} object into hex form.
+      const hex = (k: string): string => (value || { r: 0, g: 0, b: 0 })[k].toString(16).padStart(2, "0");
+      shadowValue = [BIS.COLOR_PICKER_PRIMITIVE, "#" + hex("r") + hex("g") + hex("b")];
+    } else if (typeof primitiveOrOpCode === "number") {
+      // Primitive shadow, can be stored in compressed form.
+      shadowValue = [primitiveOrOpCode, value];
+    } else {
+      // Note: Only 1-field shadow blocks are supported.
+      const shadowOpCode = primitiveOrOpCode;
+      const fieldEntries = sb3.fieldTypeMap[shadowOpCode];
+      if (fieldEntries) {
+        const fieldKey = Object.keys(fieldEntries)[0];
+        const fields = { [fieldKey]: [value] };
 
-function extractCostumes(target: Target): sb3.Costume[] {
-  return target.costumes.map(costume => {
-    const sb3Costume: sb3.Costume = {
-      assetId: costume.md5,
-      name: costume.name,
-      md5ext: `${costume.md5}.${costume.ext}`,
-      dataFormat: costume.ext,
-      rotationCenterX: costume.centerX / 2,
-      rotationCenterY: costume.centerY / 2
+        const id = generateId();
+
+        blockData[id] = {
+          opcode: shadowOpCode,
+
+          next: null,
+          parent: parentId,
+
+          fields,
+          inputs: {},
+
+          shadow: true,
+          topLevel: false
+        };
+
+        shadowValue = id;
+      }
+    }
+
+    return { shadowValue };
```

Comment:
Now that this only returns one value, you don't need to structure/destructure it